scriptworker/cot/verify.py

```diff
@@ -435,36 +435,54 @@ def verify_docker_image_sha(chain, link):
+def find_sorted_task_dependencies(task, task_name, task_id):
     """Find the taskIds of the chain of trust dependencies of a given task.
 
     Args:
         task (dict): the task definition to inspect.
-        name (str): the name of the task, for logging and naming children.
+        task_name (str): the name of the task, for logging and naming children.
         task_id (str): the taskId of the task.
 
     Returns:
-        dict: mapping dependent task ``name`` to dependent task ``taskId``.
+        list: tuples associating dependent task ``name`` to dependent task ``taskId``.
     """
-    log.info("find_task_dependencies {} {}".format(name, task_id))
+    log.info("find_sorted_task_dependencies {} {}".format(task_name, task_id))
+
+    cot_input_dependencies = [
+        _craft_dependency_tuple(task_name, task_type, task_id)
+        for task_type, task_id in task['extra'].get('chainOfTrust', {}).get('inputs', {}).items()
+    ]
+
+    upstream_artifacts_dependencies = [
+        _craft_dependency_tuple(task_name, artifact_dict['taskType'], artifact_dict['taskId'])
+        for artifact_dict in task.get('payload', {}).get('upstreamArtifacts', [])
+    ]
+
+    dependencies = [*cot_input_dependencies, *upstream_artifacts_dependencies]
+    dependencies = _filter_out_duplicates(dependencies)
+    dependencies = _sort_dependencies_by_name_then_task_id(dependencies)
+
     decision_task_id = get_decision_task_id(task)
-    decision_key = '{}:decision'.format(name)
-    task_type = guess_task_type(name)
-    dep_dict = {}
-    for key, val in task['extra'].get('chainOfTrust', {}).get('inputs', {}).items():
-        dep_dict['{}:{}'.format(name, key)] = val
-    if 'upstreamArtifacts' in task['payload']:
-        upstream_ids = {}
-        for artifact_dict in task['payload']['upstreamArtifacts']:
-            if artifact_dict['taskId'] not in upstream_ids:
-                upstream_ids[artifact_dict['taskId']] = artifact_dict['taskType']
-        for upstream_task_id, upstream_task_type in upstream_ids.items():
-            dep_dict['{}:{}'.format(name, upstream_task_type)] = upstream_task_id
+    task_type = guess_task_type(task_name)
     if decision_task_id != task_id and task_type != 'decision':
-        dep_dict[decision_key] = decision_task_id
-    log.info(dep_dict)
-    return dep_dict
+        # make sure we deal with the decision task first, or we may populate
+        # signing:build0:decision before signing:decision
+        dependencies.insert(0, _craft_dependency_tuple(task_name, 'decision', decision_task_id))
+
+    log.info('found dependencies: {}', dependencies)
+    return dependencies
+
+
+def _craft_dependency_tuple(task_name, task_type, task_id):
+    return ('{}:{}'.format(task_name, task_type), task_id)
+
+
+def _filter_out_duplicates(_list):
+    return list(set(_list))
+
+
+def _sort_dependencies_by_name_then_task_id(dependencies):
+    return sorted(dependencies, key=lambda dep: '{}_{}'.format(dep[0], dep[1]))
```

Comment:
This seems to work well.

I also tried

`<venv_path>/bin/verify_cot --task-type balrog -- bb0nJhGSTlKlMkqB8Gquwg`

and

`<venv_path>/bin/verify_cot --task-type beetmover -- B1P8M_veRm-XDhyQe6aWxQ`

and those look good as well.

The name is both to find the type of task, and for human logging.  Once you get further, maybe try running `verify_cot` against your multi-signing pushapk task.  If it helps legibility, we might want to dynamically call one `signing2`, but it isn't a requirement now.